R/layers2traces.R

```diff
@@ -1125,3 +1184,41 @@ lty2dash <- function(x) {
   }
   as.character(x)
 }
+
+# attach a new column (hovertext) to each layer of data
+# only aes given in aes_list are included in hovertext
+make_hovertext <- function(data, params, aes_list) {
+  data$hovertext <- NULL
+  hoverTextAes <- list()
+  hoverTextAes <- list(
+    sapply(aes_list, function(x) append(hoverTextAes, x = params$hoverTextAes[[x]]))
+  )
+  data <- list(data)
+  # Copy from layers2traces
+  data <- Map(function(x, y) {
+    if (nrow(x) == 0) return(x)
+    # make sure the relevant aes exists in the data
+    for (i in seq_along(y)) {
+      aesName <- names(y)[[i]]
+      if (!aesName %in% names(x)) next
+      # TODO: should we be getting the name from scale_*(name) first?
+      varName <- y[[i]]
+      # "automatically" generated group aes is not informative
+      if (identical("group", unique(varName, aesName))) next
+      # add a line break if hovertext already exists
+      if ("hovertext" %in% names(x)) x$hovertext <- paste0(x$hovertext, br())
+      # text aestheic should be taken verbatim (for custom tooltips)
+      prefix <- if (identical(aesName, "text")) "" else paste0(varName, ": ")
+      # look for the domain, if that's not found, provide the range (useful for identity scales)
+      txt <- x[[paste0(aesName, "_plotlyDomain")]] %||% x[[aesName]]
+      suffix <- tryNULL(format(txt, justify = "none")) %||% ""
+      if (inherits(x, "GeomBar") && identical(aesName, "y")) {
+        suffix <- format(x[["ymax"]] - x[["ymin"]], justify = "none")
+      }
+      x$hovertext <- paste0(x$hovertext, prefix, suffix)
+    }
+    x$hovertext <- x$hovertext %||% ""
+    x
+  }, data, hoverTextAes)
```

Comment:
Instead of copy/pasting this from the body of `layers2traces()` let's extract this bit into it's own function that we can use it both places